### PR TITLE
2.x Fix bug when Timber\Theme encoding breaks Timber caching

### PR DESCRIPTION
**Pull request by gchtr**
Resolves:

- #2429
- #2646

## Issue

In #2264, a change was added that adds a recursion for the `Timber\Theme` object. When serializing the object with `json_encode()`, this will cause the serialization to fail. Because of this, Timber can’t create a caching key and returns the same for different posts (see #2429 and #2646).

## Solution

This pull request:

- Implements the `JsonSerializable` interface for the `Timber\Theme` class so that it doesn’t break `json_encode()`.
- Adds a test for the serialization.
- Adds a check whether `json_encode()` was successfull and skips the cache if no cache key can be generated.
- Fixes testing some other theme functions (in https://github.com/timber/timber/commit/11f601e391a703edc4a4a844cb68c239db31e9b6, a change was added that caused a lot of theme tests to be skipped).

## Impact

None.

## Usage Changes

None.

## Considerations

I chose to go with `JsonSerializable` because this is a pattern that we already advise to use in cases like these: https://timber.github.io/docs/v2/guides/posts/#serialization.

This change may also be helpful for people who need theme data in a JSON-friendly version.

## Testing

Yes.

### CODE REVIEW

File: src/Theme.php [hunk 1]
```diff
@@ -25,7 +25,7 @@
  * <script src="http://example.org/wp-content/themes/my-theme/static/js/all.js"></script>
  * ```
  */
-class Theme extends Core
+class Theme extends Core implements \JsonSerializable
```
Comment (nlemoine) [inline src/Theme.php:28]: `use JsonSerializable;`

Since we're making use of `use Something` elsewhere, let's be consistent and always choose imports.

**Review by nlemoine — CHANGES_REQUESTED** [review]

File: src/Theme.php (hunk 1 above)
Comment (gchtr) [reply src/Theme.php]: 👍 Added in [`ec534a1` (#2675)](https://github.com/timber/timber/pull/2675/commits/ec534a1cf262181d3baca22d6c47b393ebdb9452).

**Comment by gchtr** [comment]
Tests only fail when running with coverage, but I couldn’t see why yet. I would merge this in and then check whether tests still fail on the 2.x branch.

**Comment by gchtr** [comment]
I fixed the tests. Test only fail on PHP 8.2, which we ignore for now. I’m merging this.

**Pull request merged**